### PR TITLE
Expression variable for the currently rendered part

### PR DESCRIPTION
**Pull request by m-kuhn**
During rendering, two new variables will be available:
- `geometry_part_count`
- `geometry_part_num` (1-based index)

Useful to apply different styles to different parts of multipart features

### CODE REVIEW

**Comment by nirvn** [comment]
@m-kuhn, I'm wondering whether the variable name should be a tiny bit less generic to avoid collision with user variables. Maybe @geometry_part_num , @feature_part_num ?

**Comment by m-kuhn** [comment]
@nirvn better?

**Comment by nirvn** [comment]
@m-kuhn yep, IMO much clearer that way.

File: src/core/qgsexpressioncontext.h [hunk 1]
```diff
@@ -378,6 +378,11 @@ class CORE_EXPORT QgsExpressionContext
      */
     void appendScope( QgsExpressionContextScope* scope );
 
+    /**
+     * Remove the last scope from the expression context.
+     */
+    void popScope();
```
Comment (nyalldawson) [inline src/core/qgsexpressioncontext.h]: Can you add a unit test for this? QgsExpressionContext has good test coverage, I don't want this to drop.

File: src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp [hunk 2]
```diff
@@ -186,17 +186,22 @@ bool QgsGeometryGeneratorSymbolLayerV2::isCompatibleWithSymbol( QgsSymbolV2* sym
...
-
 void QgsGeometryGeneratorSymbolLayerV2::render( QgsSymbolV2RenderContext& context )
 {
-  QgsGeometry geom = mExpression->evaluate( &context.renderContext().expressionContext() ).value<QgsGeometry>();
   if ( context.feature() )
   {
-    QgsFeature f = *context.feature();
+    QgsExpressionContext& expressionContext = context.renderContext().expressionContext();
+
+    QgsFeature f = expressionContext.feature();
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp]: The original `QgsFeature f = *context.feature();` would be faster, is there any particular reason for this change?

File: src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp [hunk 3]
```diff
@@ -186,17 +186,22 @@ bool QgsGeometryGeneratorSymbolLayerV2::isCompatibleWithSymbol( QgsSymbolV2* sym
...
+    QgsGeometry geom = mExpression->evaluate( &expressionContext ).value<QgsGeometry>();
     f.setGeometry( geom );
 
+    QgsExpressionContextScope* subSymbolExpressionContextScope = mSymbol->symbolRenderContext()->expressionContextScope();
+
+    subSymbolExpressionContextScope->setFeature( f );
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp]: This shouldn't be needed - if the subSymbolExpressionContextScope does not have a feature set, expressionContext will fall back to using the already set feature (which must have already been done, or line 195 would not work).

**Comment by nyalldawson** [comment]
@m-kuhn looks good to me, with the exceptions noted above. I'd also love to see a unit test to make sure the part number variable is being set correctly (eg data defined fill based on part number), as this is the kind of thing in symbology which we have a big history of regressions with ;)

File: src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp]: The previous version did not replace the geometry in the expression scope but only passed it to the symbol render function so there may have been unwanted side-effects.

Now we leave the original feature untouched (in a less-important scope) and add a new scope with a modified symbol version that hides the other one. This seems much cleaner.

File: src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp (hunk 3 above)
Comment (m-kuhn) [inline src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp]: See above. This feature should hide the one with the unmodified feature/geometry (which we need to preserve for the parent symbol).

**Comment by m-kuhn** [comment]
Tests implemented. Are you ok with my comments regarding your questions?

File: src/core/symbology-ng/qgssymbolv2.cpp [hunk 4]
```diff
@@ -916,6 +944,16 @@ QgsSymbolV2RenderContext& QgsSymbolV2RenderContext::operator=( const QgsSymbolV2
...
 }
 
+QgsExpressionContextScope* QgsSymbolV2RenderContext::expressionContextScope()
+{
+  return mExpressionContextScope;
+}
+
+void QgsSymbolV2RenderContext::setExpressionContextScope( QgsExpressionContextScope* contextScope )
+{
+  mExpressionContextScope = contextScope;
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.cpp]: This will leak - needs to delete the old scope

File: src/core/qgsexpressioncontext.cpp [hunk 5]
```diff
@@ -382,6 +382,11 @@ void QgsExpressionContext::appendScope( QgsExpressionContextScope* scope )
   mStack.append( scope );
 }
 
+void QgsExpressionContext::popScope()
+{
+  mStack.pop_back();
```
Comment (nyalldawson) [inline src/core/qgsexpressioncontext.cpp]: Better make sure this isn't empty first - I think it will crash otherwise

**Comment by nyalldawson** [comment]
Ok, I'm +1. A few little changes needed before merge:
- fix leak noted above
- fix potential crash with popScope
- add unit tests for popScope

Neat feature BTW. I'd love to get something like this in labeling too, but looking at it that's non trivial (part split happens in pal, would need to occur before registering features with pal)

**Comment by m-kuhn** [comment]
Merged in 4e9afce  

Thanks a lot for the review  @nyalldawson

**Pull request closed without merging**